ContactMechanics/Systems.py

```diff
@@ -578,10 +585,47 @@ def primal_hessian_product(self, gap):
         """Returns the hessian product of the primal_objective function.
         """
         res = self.substrate.nb_domain_grid_pts
+        # TODO: this is not parallelized yet
 
+        # TODO: do all the minimizers want to use 1d arrays ?
+        # If not the reshape(-1) has to be put there depending on the shape of
+        # gap or simple .reshape(gap.shape)
         hessp = -self.substrate.evaluate_force(gap.reshape(res)).reshape(-1)
         return hessp
 
+    def primal_minimize_proxy(self, offset, solver=bugnicourt_cg,
+                              **kwargs):
+        """
+        Convenience function. Eliminates boilerplate code for PRIMAL
+        minimisation
+        problems by encapsulating the use of constrained minimisation.
+
+        Parameters:
+        offset     -- determines indentation depth
+        disp0      -- initial guess for surface displacement. If not set, zero
+                      displacement of shape
+                      self.substrate.nb_domain_grid_pts is used
+        maxiter    -- maximum number of iterations allowed for convergence
+        logger     -- optional logger, to be used with a logger from
+                      PyCo.Tools.Logger
+        """
+        # pylint: disable=arguments-differ
+        self.disp = None
+        self.force = None
+        self.contact_zone = None
+        result = solver.constrained_conjugate_gradients(
+            self.primal_objective(offset, gradient=True),
+            self.primal_hessian_product,
+            **kwargs)
+        if result.success:
+            self.offset = offset
+            self.disp = result.x
```

Comment:
here as well this is the gap. 

I think what you should do is set self.displacement and self.gap. THat would be the most convenient.